torch/distributed/distributed_c10d.py

```diff
@@ -1526,6 +1551,83 @@ def _set_pg_timeout(timeout: timedelta, group: Optional[ProcessGroup] = None) ->
         backend._set_default_timeout(timeout)
 
 
+@_exception_logger
+@_time_logger
+def init(
+    world_size: int = -1,
+    rank: int = -1,
+    store: Optional[Store] = None,
+) -> None:
+    """
+    Initialize a distributed environment for current job. Once this API is
+    called, `get_world_size()` is guaranteed to return the number of processes
+    in current job, and `get_rank()` is guaranteed to return the rank of current
+    process in the job.
+
+    There are two main ways to initialize the distributed environment:
+        1. Specify ``store``, ``rank``, and ``world_size`` explicitly.
+        2. Infer from environment variables set by the launcher of the job (e.g.
+        torchrun, SLURM).
+
+    If using method 2, the launcher must specify the following environment variables:
+        - `RANK`: the rank of current process in the job;
+        - `WORLD_SIZE`: the number of processes in the job;
+        - `MASTER_ADDRESS`: an IP address accessible by all processes as rendezvous point;
+        - `MASTER_PORT`: a port of the above IP address over which the rendezvous happens.
+    `RANK` must be unique per process, while `WORLD_SIZE`, `MASTER_ADDRESS` and
+    `MASTER_PORT` must be the same across all processes.
+
+    Args:
+        world_size (int, optional):
+            Number of processes participating in the job. Required if ``store``
+            is specified.
+        rank (int, optional):
+            Rank of the current process (a number between 0 and
+            ``world_size``-1).  Required if ``store`` is specified.
+        store(Store, optional):
+            Key/value store accessible to all workers, used to exchange
+            connection/address information.  Mutually exclusive with environment
+            variable based initialization.
+    """
+
+    global _world
+    if _world is not None:
+        warnings.warn(
+            "dist.init() has already been called. "
+            "Re-initializing the world is currently experimental."
+        )
+
+    set_pytorch_distributed_envs_from_justknobs()
+
+    # Depending on the import order, some trace_rules functions may be evaluated
+    # during the import phase. In such a case, these functions may not correctly
+    # add the distributed related rules due to import circular dependency.
+    # We need to clear the lru_cache during the runtime to ensure the correctness
+    # of these trace_rules.
+    #
+    # Since this API must be called before all distributed code being compiled,
+    # clearing the cache here should be safe.
+    if "torch._dynamo" in sys.modules:
+        torch._dynamo.trace_rules.clear_lru_cache()
+
+    if store is None:
+        # Create one from environment variables
+        rendezvous_iterator = rendezvous(
+            "env://",
+            rank,
+            world_size,
+            # TODO: should we support configurable timeout for rendezvous?
+        )
+        store, rank, world_size = next(rendezvous_iterator)
```

Comment:
this is interesting -- we may not actually want to create a global store for performance reasons. tcpstore is fairly cheap but still hmm

we don't actually need it for get_rank()/get_world_size()

Reply:
Can we pass in the store so folks can use other stuff for example zues for rendezvous? Not sure...